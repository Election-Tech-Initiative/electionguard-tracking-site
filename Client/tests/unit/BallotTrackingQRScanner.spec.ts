import { mount } from '@vue/test-utils';

import BallotTrackingQrScanner from '@/components/BallotTrackingQrScanner.vue';

jest.mock('vue-qrcode-reader', () => ({ QrcodeCapture: jest.fn(), QrcodeStream: jest.fn() }));
jest.mock('@/components/LoadingIndicator.vue', () => ({ default: jest.fn() }));

describe('BallotTrackingQrScanner.vue', () => {
    const mockedRouterPush = jest.fn();

    const getMountedWrapper = () => mount(BallotTrackingQrScanner, {
        mocks: {
            $router: {
                push: mockedRouterPush,
            },
            $t: (msg: string): string => msg
        },
        stubs: [
            'LoadingIndicator',
            'QrcodeCapture',
            'QrcodeStream',
        ],
    });

    it('renders', () => {
        const wrapper = getMountedWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with openModal', () => {
        const wrapper = getMountedWrapper();
        wrapper.find('button').trigger('click');
        expect(wrapper.element).toMatchSnapshot('with openModal');
    });

    it('renders error when detection fails', async () => {
        const wrapper = getMountedWrapper();
        (wrapper.vm as any).onDetect(Promise.reject(new Error('error')));
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot('with error when detection fails');
    });

    it('renders error when result does\'t start with ${window.location.origin}/track/', async () => {
        const wrapper = getMountedWrapper();
        (wrapper.vm as any).onDetect(Promise.resolve({ content: 'invalid result' }));
        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot('with error when result invalid');
    });

    it('routes to tracking url when result starts with ${window.location.origin}/track', async () => {
        const wrapper = getMountedWrapper();
        const expectedSubUrl = 'track/mocked-url';
        (wrapper.vm as any).onDetect(Promise.resolve({ content: `${window.location.origin}/${expectedSubUrl}` }));
        await wrapper.vm.$nextTick();
        expect(mockedRouterPush).toBeCalledWith(expectedSubUrl);
    });
});
