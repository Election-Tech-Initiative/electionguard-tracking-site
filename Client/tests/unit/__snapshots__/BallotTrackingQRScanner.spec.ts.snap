// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BallotTrackingQrScanner.vue renders 1`] = `
<div
  id="qr"
>
  <label
    class="label"
  >
    scanQrCodeLabel
  </label>
   
  <div
    class="control"
  >
    <qrcodecapture-stub
      style="display: none;"
    />
     
    <button
      class="button is-large is-success"
    >
      
      scanQrCode
    
    </button>
  </div>
   
  <p
    class="help is-danger is-hidden"
  >
    <i
      class="fas fa-exclamation-triangle"
    />
     
  </p>
   
  <!---->
</div>
`;

exports[`BallotTrackingQrScanner.vue renders error when detection fails: with error when detection fails 1`] = `
<div
  id="qr"
>
  <label
    class="label"
  >
    scanQrCodeLabel
  </label>
   
  <div
    class="control"
  >
    <qrcodecapture-stub
      style="display: none;"
    />
     
    <button
      class="button is-large is-success"
    >
      
      scanQrCode
    
    </button>
  </div>
   
  <p
    class="help is-danger"
  >
    <i
      class="fas fa-exclamation-triangle"
    />
     qrError
  </p>
   
  <!---->
</div>
`;

exports[`BallotTrackingQrScanner.vue renders error when result does't start with \${window.location.origin}/track/: with error when result invalid 1`] = `
<div
  id="qr"
>
  <label
    class="label"
  >
    scanQrCodeLabel
  </label>
   
  <div
    class="control"
  >
    <qrcodecapture-stub
      style="display: none;"
    />
     
    <button
      class="button is-large is-success"
    >
      
      scanQrCode
    
    </button>
  </div>
   
  <p
    class="help is-danger"
  >
    <i
      class="fas fa-exclamation-triangle"
    />
     qrError
  </p>
   
  <!---->
</div>
`;

exports[`BallotTrackingQrScanner.vue renders with openModal: with openModal 1`] = `
<div
  id="qr"
>
  <label
    class="label"
  >
    scanQrCodeLabel
  </label>
   
  <div
    class="control"
  >
    <qrcodecapture-stub
      style="display: none;"
    />
     
    <button
      class="button is-large is-loading is-info"
      disabled="disabled"
    >
      
      scanQrCode
    
    </button>
  </div>
   
  <p
    class="help is-danger is-hidden"
  >
    <i
      class="fas fa-exclamation-triangle"
    />
     
  </p>
   
  <div
    class="modal is-active"
  >
    <div
      class="modal-background"
    />
     
    <div
      class="modal-content"
    >
      <div
        class="box has-text-centered"
      >
        <button
          aria-label="close"
          class="delete is-pulled-right"
        />
         
        <h6
          class="title is-5"
        >
          scanQrCode
        </h6>
         
        <h6
          class="subtitle is-7"
        >
          scanQrCodeHelpMessage
        </h6>
         
        <div
          class="content"
        >
          <loadingindicator-stub />
           
          <qrcodestream-stub />
        </div>
      </div>
    </div>
  </div>
</div>
`;
